types/Connection.d.ts

```diff
@@ -0,0 +1,212 @@
+      asPromise(): Promise<this>;
+
+      /** Closes the connection */
+      close(force: boolean, callback: CallbackWithoutResult): void;
+      close(callback: CallbackWithoutResult): void;
+      close(force?: boolean): Promise<void>;
+
+      /** Retrieves a collection, creating it if not cached. */
+      collection<T extends AnyObject = AnyObject>(name: string, options?: mongodb.CreateCollectionOptions): Collection<T>;
+
+      /** A hash of the collections associated with this connection */
+      readonly collections: { [index: string]: Collection };
+
+      /** A hash of the global options that are associated with this connection */
+      readonly config: any;
+
+      /** The mongodb.Db instance, set when the connection is opened */
+      readonly db: mongodb.Db;
+
+      /**
+         * Helper for `createCollection()`. Will explicitly create the given collection
+         * with specified options. Used to create [capped collections](https://docs.mongodb.com/manual/core/capped-collections/)
+         * and [views](https://docs.mongodb.com/manual/core/views/) from mongoose.
+         */
+      createCollection<T extends AnyObject = AnyObject>(name: string, options: mongodb.CreateCollectionOptions, callback: Callback<mongodb.Collection<T>>): void;
+      createCollection<T extends AnyObject = AnyObject>(name: string, callback: Callback<mongodb.Collection<T>>): void;
+      createCollection<T extends AnyObject = AnyObject>(name: string, options?: mongodb.CreateCollectionOptions): Promise<mongodb.Collection<T>>;
+
+      /**
+         * Removes the model named `name` from this connection, if it exists. You can
+         * use this function to clean up any models you created in your tests to
+         * prevent OverwriteModelErrors.
+         */
+      deleteModel(name: string): this;
+
+      /**
+         * Helper for `dropCollection()`. Will delete the given collection, including
+         * all documents and indexes.
+         */
+      dropCollection(collection: string, cb: CallbackWithoutResult): void;
+      dropCollection(collection: string): Promise<void>;
+
+      /**
+         * Helper for `dropDatabase()`. Deletes the given database, including all
+         * collections, documents, and indexes.
+         */
+      dropDatabase(cb: CallbackWithoutResult): void;
+      dropDatabase(): Promise<void>;
+
+      /** Gets the value of the option `key`. */
+      get(key: string): any;
+
+      /**
+         * Returns the [MongoDB driver `MongoClient`](http://mongodb.github.io/node-mongodb-native/3.5/api/MongoClient.html) instance
+         * that this connection uses to talk to MongoDB.
+         */
+      getClient(): mongodb.MongoClient;
+
+      /**
+         * The host name portion of the URI. If multiple hosts, such as a replica set,
+         * this will contain the first host name in the URI
+         */
+      readonly host: string;
```

Comment:
please give me feedback regarding the readonly on host, user, pass etc. 
Is it wrong? Is there actually a use case for having them writeable?

If I am wrong, then I would like to fix it before we have regression reports

Reply:
I think `readonly` should be fine, modifying `host` is a noop.